core/arch/emit_utils_shared.c

```diff
@@ -5534,6 +5536,56 @@ emit_special_ibl_xfer(dcontext_t *dcontext, byte *pc, generated_code_t *code, ui
         in = instrlist_first(custom_ilist);
     }
 
+#ifdef UNIX
+    /* i#4670: Jump to the unlinked IBL target if there are pending signals. This is
+     * required to bound delivery time for signals received while executing fragments
+     * that use the special ibl xfer trampoline, which uses a different (un)linking
+     * mechanism.
+     */
+    insert_shared_get_dcontext(dcontext, &ilist, NULL, true);
+#    ifdef X86
+    APP(&ilist,
+        XINST_CREATE_cmp(dcontext,
+                         OPND_DC_FIELD(false, dcontext, OPSZ_1, SIGPENDING_OFFSET),
+                         OPND_CREATE_INT8(0)));
+    insert_shared_restore_dcontext_reg(dcontext, &ilist, NULL);
+    APP(&ilist,
+        XINST_CREATE_jump_cond(dcontext, DR_PRED_GT, opnd_create_pc(ibl_unlinked_tgt)));
+#    elif defined(AARCHXX)
+    instr_t *skip_unlinked_tgt_jump = INSTR_CREATE_label(dcontext);
+    /* Reusing SCRATCH_REG5 which contains dcontext currently. */
+    APP(&ilist,
+        INSTR_CREATE_ldrsb(dcontext, opnd_create_reg(SCRATCH_REG5),
+                           OPND_DC_FIELD(false, dcontext, OPSZ_1, SIGPENDING_OFFSET)));
```

Comment:
Another option is that we add a new member to `dcontext_t` for `is_thread_unlinking` which can be incremented in all scenarios requiring proactive unlinking-relinking of fragments. But I'm aware of only the pending signals scenario right now for which `signals_pending>0` works just as well.

Reply:
I think long-term #4804 is the cleanest solution and we shouldn't anticipate having any other fastpath checks for unlinking: we should use regular link/unlink paths for any other cases.